File: no_rec.js
function caminhoDasPedras(caminho) {
	const table = Array.from({ length: caminho.length }, () => [0, 0]);
	table[0] = [1, 1];

	for (let i = 1; i < caminho.length; i++) {
		if (caminho[i] === "0") {
			continue;
		}

		let res = table[i - 1][0];
		if (i >= 2) {
			res += table[i - 2][0];
		}

		table[i][1] = res;

		if (i >= 3) {
			res += table[i - 3][1];
		}

		table[i][0] = res;
	}

	return table[caminho.length - 1][0];
}

const args = process.argv.slice(2);

if (args.length != 1) {
	console.log("Uso: node no_rec.js <caminho>");
	process.exit(1);
}

console.log(caminhoDasPedras(args[0]));

File: src/bin/no_rec.rs
use std::collections::HashMap;

fn caminho_das_pedras(caminho: &str) -> usize {
    let mut table = HashMap::with_capacity(caminho.len());

    table.insert((0, false), 1);
    table.insert((0, true), 1);

    for (i, ch) in caminho.chars().enumerate().skip(1) {
        if ch == '0' {
            continue;
        }

        let res = (1..=2).flat_map(|x| table.get(&(i - x, false))).sum();
        table.insert((i, true), res);
        table.insert((i, false), res + table.get(&(i - 3, true)).unwrap_or(&0));
    }

    *table.get(&(caminho.len() - 1, false)).unwrap_or(&0)
}

fn main() {
    let args = std::env::args().collect::<Vec<_>>();

    if args.len() != 2 {
        println!("Uso: {} <caminho>", args[0]);
        return;
    }

    println!("{}", caminho_das_pedras(&args[1]));
}
